File: cs/Problems/Day7.cs
namespace aoc24.Problems;

public sealed class Day7 : IProblem<long>
{
    public long Solve(string input) => CalculateCalibrationValuesOptimized(input);

    public static long CalculateCalibrationValuesOptimized(ReadOnlySpan<char> input)
    {
        long totalCalibration = 0;

        Span<long> numbers = stackalloc long[15];
        var lineEnumerator = input.Split(InputReader.NewLine);

        while (lineEnumerator.MoveNext())
        {
            var line = input[lineEnumerator.Current];
            int splitInx = line.IndexOf(':');
            long testValue = long.Parse(line[..splitInx]);
            int length = InputParser.ParseNumbers(numbers, line[(splitInx + 1)..], ' ');

            if (OperatorsEqualTestValue(testValue, numbers[0], numbers, length, 1))
                totalCalibration += testValue;
        }

        return totalCalibration;
    }

    public static long CalculateCalibrationValues(string input)
    {
        long totalCalibration = 0;

        foreach (var line in input.Split(InputReader.NewLine))
        {
            int splitInx = line.IndexOf(':');
            long testValue = long.Parse(line[..splitInx]);
            var numbers = line[(splitInx + 2)..].Split(" ").Select(long.Parse).ToArray();

            if (OperatorsEqualTestValue(testValue, numbers[0], numbers, numbers.Length, 1))
                totalCalibration += testValue;
        }

        return totalCalibration;
    }

    public static bool OperatorsEqualTestValue(long testValue, long currentValue, ReadOnlySpan<long> numbers, int length, int inx)
    {
        if (currentValue > testValue) // If current exceeds test value, it will never match.
            return false;

        if (inx == length) // In the last iteration, compare the values.
            return testValue == currentValue;

        long addValue = currentValue + numbers[inx];
        long multValue = currentValue * numbers[inx];

        return OperatorsEqualTestValue(testValue, addValue, numbers, length, inx + 1)
            || OperatorsEqualTestValue(testValue, multValue, numbers, length, inx + 1);
    }
}

File: go/problems/day7.go
package problems

import (
	"strconv"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day7 struct{}

func (d Day7) Solve(input string) int64 {
	var r int64 = 0
	nms := make([]int64, 0)

	for _, line := range strings.Split(input, shared.NewLine) {
		nms = nms[:0] // reset the slice...
		split := strings.Split(line, ":")

		for _, sn := range strings.Split(split[1], " ") {
			n, _ := strconv.ParseInt(sn, 10, 64)
			nms = append(nms, n)
		}

		target, _ := strconv.ParseInt(split[0], 10, 64)
		if operatorEqualsValue(target, nms[0], nms, 1) {
			r += target
		}
	}

	return r
}

func operatorEqualsValue(target, curr int64, nms []int64, inx int) bool {
	if curr > target {
		return false
	}

	// compare only in the last iteration
	if inx == len(nms) {
		return curr == target
	}

	add := curr + nms[inx]
	prod := curr * nms[inx]

	return operatorEqualsValue(target, add, nms, inx+1) || operatorEqualsValue(target, prod, nms, inx+1)
}
